File: mods/java/hytale/src/main/java/com/alaydriem/bedrockvoicechat/hytale/HytaleModConfig.java
package com.alaydriem.bedrockvoicechat.hytale;

import com.hypixel.hytale.codec.Codec;
import com.hypixel.hytale.codec.KeyedCodec;
import com.hypixel.hytale.codec.builder.BuilderCodec;

/**
 * Hytale-native config class using BuilderCodec.
 * Mirrors ModConfig fields but uses Hytale's serialization.
 */
public class HytaleModConfig {
    public static final BuilderCodec<HytaleModConfig> CODEC =
        BuilderCodec.builder(HytaleModConfig.class, HytaleModConfig::new)
            .append(
                new KeyedCodec<>("bvc-server", Codec.STRING),
                (cfg, val) -> cfg.bvcServer = val,
                cfg -> cfg.bvcServer
            ).add()
            .append(
                new KeyedCodec<>("access-token", Codec.STRING),
                (cfg, val) -> cfg.accessToken = val,
                cfg -> cfg.accessToken
            ).add()
            .append(
                new KeyedCodec<>("minimum-players", Codec.INTEGER),
                (cfg, val) -> cfg.minimumPlayers = val,
                cfg -> cfg.minimumPlayers
            ).add()
            .build();

    private String bvcServer = "";
    private String accessToken = "";
    private int minimumPlayers = 2;

    public HytaleModConfig() {}

    public String getBvcServer() {
        return bvcServer;
    }

    public String getAccessToken() {
        return accessToken;
    }

    public int getMinimumPlayers() {
        return minimumPlayers;
    }

    public boolean isValid() {
        return bvcServer != null && !bvcServer.isBlank()
            && accessToken != null && !accessToken.isBlank();
    }
}

File: mods/java/hytale/src/main/kotlin/com/alaydriem/bedrockvoicechat/hytale/HytaleModConfig.kt
package com.alaydriem.bedrockvoicechat.hytale

import com.hypixel.hytale.codec.Codec
import com.hypixel.hytale.codec.KeyedCodec
import com.hypixel.hytale.codec.builder.BuilderCodec

/**
 * Hytale-native config class using BuilderCodec.
 * Mirrors ModConfig fields but uses Hytale's serialization.
 */
class HytaleModConfig {
    var bvcServer: String = ""
    var accessToken: String = ""
    var minimumPlayers: Int = 2

    fun isValid(): Boolean =
        bvcServer.isNotBlank() && accessToken.isNotBlank()

    companion object {
        /**
         * Lazily initialized codec to avoid ClassNotFoundException during testing.
         * Only initialized when actually needed (at plugin runtime).
         */
        @JvmStatic
        val CODEC: BuilderCodec<HytaleModConfig> by lazy {
            BuilderCodec.builder(
                HytaleModConfig::class.java,
                ::HytaleModConfig
            )
                .append(
                    KeyedCodec("bvc-server", Codec.STRING),
                    { cfg, value -> cfg.bvcServer = value },
                    { cfg -> cfg.bvcServer }
                ).add()
                .append(
                    KeyedCodec("access-token", Codec.STRING),
                    { cfg, value -> cfg.accessToken = value },
                    { cfg -> cfg.accessToken }
                ).add()
                .append(
                    KeyedCodec("minimum-players", Codec.INTEGER),
                    { cfg, value -> cfg.minimumPlayers = value },
                    { cfg -> cfg.minimumPlayers }
                ).add()
                .build()
        }
    }
}
